### ISSUE TITLE
Type Hierarchy Simplification

### ISSUE THREAD

**Issue by ccoffrin** [lanl-ansi/PowerModels.jl#192]
Look into using the [def macro](https://discourse.julialang.org/t/def-macro-generator-broken-on-master/1096/3) or https://github.com/tbreloff/ConcreteAbstractions.jl to simplify the type hierarchy.

**Comment by ccoffrin** [comment]
Since ConcreteAbstractions is not registered as a Julia package, it seems that the def macro approach will be best option at this time.  Archive from discourse,
```
macro def(name, definition)
    return quote
        macro $(esc(name))()
            esc($(Expr(:quote, definition)))
        end
    end
end

@def test begin
    a = 2
end
```

**Comment by ccoffrin** [comment]
`AbstractDCPForm` is a good candidate for renaming.  The primary semantics are active power only with `va` with `vm=1.0`.

**Comment by ccoffrin** [comment]
Another idea to explore. Add getter and setter functions for the parts of GenericPowerModel, abstract the code to only use these accessor functions.  Then package extensions can bring their own type as long, as it implements the accessor API.  This may also provide a clear path for moving more generic code into InfrastructureModels.jl.

**Comment by claytonpbarrows** [comment]
Also, ReusePatterns.jl can help forward accessor functions and reduce code duplication.

**Comment by ccoffrin** [comment]
closed in #530 / v0.13
